<?xml version="1.0"?>

<!-- Make the namespace of the robot-->
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="rrr_arm">

<xacro:include filename="$(find drone_simulation)/urdf/Kwad.xacro"/>
<include file="$(find drone_simulation)/urdf/Kwad.xacro"/>

    <!--Define constants-->
    <xacro:property name="link_type" value="cuboidal" />
    <xacro:property name="M_PI" value="3.14159"/>
    <xacro:property name="mass_1" value="0.050" />
    <xacro:property name="mass_2" value="0.067184" />
    <xacro:property name="mass_3" value="0.081567" />
    <xacro:property name="mass_4" value="0.058868" />


    <xacro:property name="radius_1" value="0.1" />  <!--base_link radius-->
    <xacro:property name="radius_2" value="0.012" /> <!--link_1 radius-->
    <xacro:property name="radius_3" value="0.012" /><!--link_2 radius-->
    <xacro:property name="radius_4" value="0.012" /><!--link_3 radius-->

    <xacro:property name="length_1" value="0.05" /> <!--base_link length-->
    <xacro:property name="length_2" value="0.05" />  <!--link_1 length-->
    <xacro:property name="length_3" value="0.15" />  <!--link_2 length-->
    <xacro:property name="length_4" value="0.10" /> <!--link_3 length-->

    <xacro:property name="width_1"  value="0.2"/>
    <xacro:property name="height_1" value="0.02"/>

    <!--Define Rviz colors-->
    <material name="Black">
        <color rgba="0.0 0.0 0.0 1.0"/>
    </material>
    <material name="Red">
        <color rgba="0.8 0.0 0.0 1.0"/>
    </material>
    <material name="Blue">
        <color rgba="0.0 0.0 1.0 1.0"/>
    </material>
    <material name="LightBlue">
        <color rgba="0.589 0.793 1.0 1.0"/>
    </material>
    <material name="Grey">
        <color rgba="0.5 0.5 0.5 1.0"/>
    </material>
    <material name="White">
        <color rgba="1.0 1.0 1.0 1.0"/>
    </material>

    <!-- macros for inertia -->
    <xacro:macro name="inertial_block" params="mass length radius">
        <inertial>
        <origin xyz="0 0 ${length/2}" rpy="0 0 0"/>
        <mass value="${mass}"/>
        <inertia
            ixx="${mass / 12.0 * (2*radius*2*radius + length*length)}" ixy="0.0" ixz="0.0"
            iyy="${mass / 12.0 * (length*length + 2*radius*2*radius)}" iyz="0.0"
            izz="${mass / 12.0 * (2*radius*2*radius + 2*radius*2*radius)}" />
    </inertial>
    </xacro:macro>

    <!-- macros for transmission -->
    <xacro:macro name="transmission_block" params="joint_name">
	  <transmission name="tran1">
	    <type>transmission_interface/SimpleTransmission</type>
	    <joint name="${joint_name}">
	      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    </joint>
	    <actuator name="motor1">
	      <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	      <mechanicalReduction>1</mechanicalReduction>
	    </actuator>
	  </transmission>
   </xacro:macro>
    

    <!--Define Gazebo colors-->
    <xacro:macro name="gazebo_grey_link_color_block" params="link_name">
    <gazebo reference="${link_name}">
    <visual>
      <material>
        <ambient>0.5 0.5 0.5 1.0</ambient>
        <diffuse>0.56666695 0.56666695 0.56666695 1.0</diffuse>
        <specular>0.6000003 0.6000003 0.6000003 1.0</specular>
      </material>
    </visual>
    </gazebo>
  </xacro:macro>

  <xacro:macro name="gazebo_blue_link_color_block" params="link_name">
    <gazebo reference="${link_name}">
    <visual>
      <material>
        <ambient>0.589 0.793 1.0 1.0</ambient>
        <diffuse>0.56666695 0.56666695 0.56666695 1.0</diffuse>
        <specular>0.6000003 0.6000003 0.6000003 1.0</specular>
      </material>
    </visual>
    </gazebo>
  </xacro:macro>
  
    <!--=============================================================================-->
    <!--world link-->

    <!-- Attach Quadrotor to RRR Arm ================================================-->
    <joint name="quadrotor_joint" type="fixed">
        <parent link="base_link1"/>
        <child link="base_link"/>
        <origin xyz="0.0 0.0 0.0" rpy="3.142 0 0"/> 
    </joint>
    
    <!--base_link-->
    <link name="base_link">
        <xacro:inertial_block mass="${mass_1}" length="${length_1}" radius="${width_1}"/>
        <visual>
            <origin xyz="0.0 0.0 0.01" rpy="0.0 0.0 0.0"/>
            <geometry>
                <!--cylinder radius="${radius_1}" length="${length_1}"/-->
                <box size="${length_1} ${width_1} ${height_1}"/>
                
            </geometry>
            <material name="LightBlue"/>
        </visual>

        <collision>
            <origin xyz="0.0 0.0 0.01" rpy="0.0 0.0 0.0"/>
            <geometry>
                <box size="${length_1} ${width_1} ${height_1}"/>
            </geometry>>
        </collision>

        
    </link>

    <xacro:gazebo_blue_link_color_block link_name="base_link"/>

    <joint name="joint_1" type="prismatic">
        <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
        <parent link="base_link"/>
        <child link="link_1"/>
        <axis xyz="0.0 1 0"/>
        <limit lower="-0.1" upper="0.1" effort="300.0" velocity="1.0"/>
    </joint>

    <!--=============================================================================-->
    <!--link_1-->
    <link name="link_1">
        <xacro:inertial_block mass="${mass_2}" length="${length_2}" radius="${radius_2}"/>
        
        <visual>
            <origin xyz="0.0 0.0 ${length_2/2.0}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="${radius_2}" length="${length_2}"/>
            </geometry>
            <material name="Grey"/>
        </visual>

        <collision>
            <origin xyz="0.0 0.0 ${length_2/2.0}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="${radius_2}" length="${length_2}"/>
            </geometry>>
        </collision>
    </link>
    
    <xacro:gazebo_grey_link_color_block link_name="link_1"/>
    <!-- <xacro:gazebo_grey_link_color_block link_name="link_1"> -->
    
    <link name="link_1_endlink">
        <xacro:inertial_block mass="${mass_2/3.0}" length="${2*radius_2+2*radius_3}" radius="${radius_2}"/>
        <visual>
            <origin xyz="0.0 0.0 ${(2*radius_2+2*radius_3)/2.0}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="${radius_2*1.1}" length="${2*radius_2+2*radius_3}"/>
            </geometry>
            <material name="LightBlue"/>
        </visual>

        <collision>
            <origin xyz="0.0 0.0 ${(2*radius_2+2*radius_3)/2.0}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="${radius_2*1.1}" length="${2*radius_2+2*radius_3}"/>
            </geometry>
        </collision>

    </link>


    <xacro:gazebo_blue_link_color_block link_name="link_1_endlink"/>



    <joint name="link_1_endjoint" type="fixed">
        <origin xyz="${-radius_2} 0.0 ${length_2 - 0.01}" rpy="0.0 ${M_PI/2.0} 0.0"/>
        <parent link="link_1"/>
        <child link="link_1_endlink"/>
    </joint>

    <gazebo reference="link_1_endjoint">
        <disableFixedJointLumping>true</disableFixedJointLumping> 
        <preserveFixedJoint>true</preserveFixedJoint>
    </gazebo>

    <joint name="joint_2" type="revolute">
        <origin xyz="${radius_2 + radius_3} 0.0 ${length_2}" rpy="0.0 0.0 0.0"/>
        <parent link="link_1"/>
        <child link="link_2"/>
        <axis xyz="1.0 0.0 0.0"/>
        <limit lower="-${M_PI / 2}" upper="${M_PI / 2}" effort="300.0" velocity="1.0"/>
    </joint>

    <!--=============================================================================-->
    <!--link_2-->
    <link name="link_2">
        <xacro:inertial_block mass="${mass_3}" length="${length_3}" radius="${radius_3}"/>
        <visual>
            <origin xyz="0.0 0.0 ${length_3/2.0}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="${radius_3}" length="${length_3}"/>
            </geometry>
            <material name="Grey"/>
        </visual>

        <collision>
            <origin xyz="0.0 0.0 ${length_3/2.0}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="${radius_3}" length="${length_3}"/>
            </geometry>>
        </collision>
    </link>

    <xacro:gazebo_grey_link_color_block link_name="link_2"/>

    <link name="link_2_endlink">
        <xacro:inertial_block mass="${mass_3/3.0}" length="${2*radius_3+2*radius_4}" radius="${radius_3}"/>
        <visual>
            <origin xyz="0.0 0.0 ${(2*radius_3+2*radius_4)/2.0}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="${radius_3*1.1}" length="${2*radius_3+2*radius_4}"/>
            </geometry>
            <material name="LightBlue"/>
        </visual>

        <collision>
            <origin xyz="0.0 0.0 ${(2*radius_3+2*radius_4)/2.0}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="${radius_3*1.1}" length="${2*radius_3+2*radius_4}"/>
            </geometry>
        </collision>
    </link>

    <xacro:gazebo_blue_link_color_block link_name="link_2_endlink"/>

    <joint name="link_2_endjoint" type="fixed">
        <origin xyz="${-radius_3} 0.0 ${length_3 - 0.01}" rpy="0.0 ${M_PI/2.0} 0.0"/>
        <parent link="link_2"/>
        <child link="link_2_endlink"/>
    </joint>

    <gazebo reference="link_2_endjoint">
        <disableFixedJointLumping>true</disableFixedJointLumping> 
        <preserveFixedJoint>true</preserveFixedJoint>
    </gazebo>

    <joint name="joint_3" type="revolute">
        <origin xyz="${radius_3 + radius_4} 0.0 ${length_3 - radius_3}" rpy="0.0 0.0 0.0"/>
        <parent link="link_2"/>
        <child link="link_3"/>
        <axis xyz="1.0 0.0 0.0"/>
        <limit lower="-${M_PI / 2}" upper="${M_PI / 2}" effort="300.0" velocity="1.0"/>
    </joint>

    <!--=============================================================================-->
    <!--link_3-->
    <link name="link_3">
        <xacro:inertial_block mass="${mass_4}" length="${length_4}" radius="${radius_4}"/>
        <visual>
            <origin xyz="0.0 0.0 ${length_4/2.0}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="${radius_4}" length="${length_4}"/>
            </geometry>
            <material name="Grey"/>
        </visual>

        <collision>
            <origin xyz="0.0 0.0 ${length_4/2.0}" rpy="0.0 0.0 0.0"/>
            <geometry>
                <cylinder radius="${radius_4}" length="${length_4}"/>
            </geometry>>
        </collision>
    </link>

    <xacro:gazebo_grey_link_color_block link_name="link_3"/>
    <!-- Transmissions for ros control -->
    <transmission name="tran1">
	  <type>transmission_interface/SimpleTransmission</type>
	  <joint name="joint_1">
	    <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	  </joint>
	  <actuator name="motor1">
	    <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    <mechanicalReduction>1</mechanicalReduction>
	  </actuator>
	</transmission>

    <transmission name="tran2">
	  <type>transmission_interface/SimpleTransmission</type>
	  <joint name="joint_2">
	    <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	  </joint>
	  <actuator name="motor2">
	    <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    <mechanicalReduction>1</mechanicalReduction>
	  </actuator>
	</transmission>

    <transmission name="tran3">
	  <type>transmission_interface/SimpleTransmission</type>
	  <joint name="joint_3">
	    <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	  </joint>
	  <actuator name="motor3">
	    <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
	    <mechanicalReduction>1</mechanicalReduction>
	  </actuator>
	</transmission>


    <!--<xacro:transmission_block joint_name="joint_1"/>
    <xacro:transmission_block joint_name="joint_2"/>
    <xacro:transmission_block joint_name="joint_3"/>
	
    <xacro:transmission_block joint_name="joint_4"/>
    
    <xacro:transmission_block joint_name="gripper_joint_1"/>
    <xacro:transmission_block joint_name="gripper_joint_2"/>-->

    <!-- ros_control plugin -->
    <gazebo>
        <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
        <robotNamespace>/rrr_arm</robotNamespace>
        </plugin>
    </gazebo>

</robot>
